Make Resto Great Again

kfinch opened the pull request:
Implemented Checklist for Resto Druid, enabling me to return the spec to 'great' status.

Also removed most of the custom Resto Druid ABC code, as it was just duplicating core functionality. We now fallback on the core version.

Also tweaked WG/rejuv ratio, as it was a little strict for a "guessing" suggestion.

Here's me healing poorly:
LOG: wowanalyzer.com/report/4hqAvJXRyg1kFPd2/3-Heroic+Portal+Keeper+Hasabel+-+Wipe+3+(7:27)/Aslann

SCREENSHOTS:
![resto-checklist-1](https://user-images.githubusercontent.com/7143486/35768937-679f0076-08d1-11e8-984e-40d54c9f121a.png)
![resto-checklist-2](https://user-images.githubusercontent.com/7143486/35768938-6beb1336-08d1-11e8-922f-84d50635a6e9.png)
![resto-checklist-3](https://user-images.githubusercontent.com/7143486/35768939-704e0b0e-08d1-11e8-8747-0257e215ec6b.png)

buimichael commented on src/Parser/Druid/Restoration/Modules/Features/Checklist.js:
```diff
@@ -0,0 +1,166 @@
+          }),
+          new GenericCastEfficiencyRequirement({
+            spell: SPELLS.INNERVATE,
+          }),
```
Should also include ToL if specced.

kfinch replied:
Good catch, I'll add that.

buimichael commented on src/Parser/Druid/Restoration/Modules/Features/Checklist.js:
```diff
@@ -0,0 +1,166 @@
+            name: <Wrapper>Don't overuse <SpellLink id={SPELLS.REGROWTH.id} icon /></Wrapper>,
+            check: () => this.clearcasting.nonCCRegrowthsSuggestionThresholds,
+            tooltip: `This is the number of no-clearcasting Regrowths you cast per minute. Regrowth is very mana inefficient, and should only be used in emergency situations (and when you've already expended Swiftmend). Usually, you should rely on other healers in your raid to triage.`,
+          }),
```
Could also include "Use all of your mana effectively" i.e. making sure the player has used up most of their mana at the end of the encounter.

buimichael replied:
I would probably move the "Used a second potion" here as it's directly tied with mana management.

kfinch replied:
Oh yeah, forgot about the mana use rate one, I'll put that in. The second potion thing is tied to a generic rule, I don't think it's worth trying to mess with that.

buimichael commented on src/Parser/Druid/Restoration/Modules/Features/Checklist.js:
```diff
@@ -0,0 +1,166 @@
+      description: <Wrapper>Restoration Druids unfortunately do not have many tools to deal with burst damage, but you should take care to use the ones you have. The below percentages represent the percentage of time you kept each spell on cooldown.</Wrapper>,
+      requirements: () => {
+        return [
+          new GenericCastEfficiencyRequirement({
```
I think this could be excluded. But if you're using t20 then it should appear under "User your healing CDs"

kfinch replied:
Yeah, got similar feedback from #resto. I'll probably exclude it entirely from this area, and add it to the cooldowns dropdown if player has SotF, SotA, or 4T20.

buimichael commented on src/Parser/Druid/Restoration/Modules/Features/Checklist.js:
```diff
@@ -0,0 +1,166 @@
+      },
+    }),
+    new PreparationRule(),
+  ];
```
Consider having  an "Avoid overhealing" section as well?

kfinch replied:
I'm kinda skeptical about that, because it's so unavoidable for much of what Resto does.

buimichael commented on src/Parser/Druid/Restoration/Modules/Features/Checklist.js:
```diff
@@ -0,0 +1,166 @@
+    }),
+    new PreparationRule(),
+  ];
+}
```
I like the "Pick the right tools for the fight" which the holy paladin checklist has. We should probably include that aswell, for ToL/Cult (sotf LUL) and legendaries. Also give "error" if they are playing prosperity without t20.

kfinch replied:
Good ideas, I wasn't sure what to do about the talent things. Probably group talent choice metrics together like that is good idea.

kfinch commented:
@buimichael Ok, pushed some updates. I kept Swiftmend in for now, but lowered the thresholds and also put a disclaimer in the description saying it should only be used in an emergency. I added Velens and ToL to the "use your cooldowns" rule. I added mana usage check to the "manage mana" rule. I added the existing talent throughput checks we have to a new "Pick the right tools for the fight" rule.

I'd like to add some more talent usage checkers, but I think that's better saved for a later PR.
